src/opam.nix

```diff
@@ -382,6 +384,100 @@ in rec {
+      # Note we are assuming no dev-repos will have different names here, but we also assume
+      # this later when we will symlink in the duniverse directory based on this name.
+      let duplicateSrc = lib.lists.findSingle (src: src.name == newSrc.name) null "multiple" srcs; in
+      # Multiple duplicates should never be found as we deduplicate on every new element.
+      assert duplicateSrc != "multiple";
+      if duplicateSrc == null then srcs ++ [ newSrc ]
+      # > If packages from the same repo were resolved to different URLs, we need to pick
+      # > a single one. Here we decided to go with the one associated with the package
+      # > that has the higher version. We need a better long term solution as this won't
+      # > play nicely with pins for instance.
+      # > The best solution here would be to use source trimming, so we can pull each individual
+      # > package to its own directory and strip out all the unrelated source code but we would
+      # > need dune to provide that feature.
+      # See [opam-monorepo](https://github.com/tarides/opam-monorepo/blob/9262e7f71d749520b7e046fbd90a4732a43866e9/lib/duniverse.ml#L143-L157)
+      else if duplicateSrc.version >= newSrc.version then srcs
+      else (lib.lists.remove duplicateSrc srcs) ++ [ newSrc ];
+    in builtins.foldl' op [ ] srcs;
+
+  mkMonorepo = srcs:
+    let
+      # derivation that fetches the source
+      mkSrc = { name, version, src }:
+        bootstrapPackages.pkgsBuildBuild.stdenv.mkDerivation ({
+          inherit name version src;
+          phases = [ "unpackPhase" "installPhase" ];
+          installPhase = ''
+            mkdir $out
+            cp -R . $out
+          '';
+        });
+    in
+    builtins.listToAttrs (builtins.map (src: nameValuePair src.name (mkSrc src)) srcs);
+
+  queryToMonorepo = { repos ? [ opamOverlays opamRepository ], resolveArgs ? { }
+      , filterPkgs ? [ ] }:
+    query:
+    pipe query [
+      # pass monorepo = 1 to pick up dependencies marked with {?monorepo}
+      (opamList (joinRepos repos) (resolveArgs // { env.monorepo=1; }))
+      opamListToQuery
+      (queryToDefs repos)
+      (defsToSrcs filterPkgs)
+      deduplicateSrcs
+      mkMonorepo
+    ];
+  
+  buildOpamMonorepo = { repos ? [ opamOverlays opamRepository ], pkgs ? bootstrapPackages
+    , resolveArgs ? { dev = true; }, pinDepends ? true, recursive ? false
+    , filterPkgs ? [ "ocaml-system" "opam-monorepo" ] }@args:
+    project: query:
+    let
+      repo = makeOpamRepo' recursive project;
+      latestVersions = mapAttrs (_: last) (listRepo repo);
+
+      pinDeps = concatLists (attrValues (mapAttrs
+        (name: version: getPinDepends repo.passthru.pkgdefs.${name}.${version})
+        latestVersions));
+    in queryToMonorepo {
+      repos = [ repo ] ++ optionals pinDepends pinDeps ++ repos;
+      # filter all query packages from monorepo
+      filterPkgs = filterPkgs ++ [ builtins.attrNames query ];
```

Comment:
`[ builtins.attrNames query ]` is a list from two elements. You probably want `[ (builtins.attrNames query) ]`

Reply:
Also, please `inherit` stuff from `builtins` or `lib` at the top of the file, and then just use the functions directly without `builtins`.

Reply:
Yep, using `(builtins.attrNames query)` fixed it! And will do.